File: JavaScript/DesignPatterns/ChainOfResponsibilityDesignPattern.js
'use strict';

class Request {
    constructor(priority = 0) {
        this.priority = priority;
    }
}

class IHandler {
    constructor() {
        this.nextHandler = null;
    }

    setNextHandler(handler) {
        this.nextHandler = handler;
        return this.nextHandler;
    }

    getNextHandler() {
        return this.nextHandler;
    }

    isAllowed(request) {
        throw new Error("Not implemented");
    }

    processRequest(request) {
        throw new Error("Not implemented");
    }
}

class Level1Handler extends IHandler {
    isAllowed(request) {
        return request.priority < 8;
    }

    processRequest(request) {
        if (this.isAllowed(request)) {
            console.log("Processing the request : lvel1 is processing");
        } else {
            this.getNextHandler().processRequest(request);
        }
    }
}

class Level2Handler extends IHandler {
    isAllowed(request) {
        return request.priority < 15;
    }

    processRequest(request) {
        if (this.isAllowed(request)) {
            console.log("Processing the request : lvel1 is processing");
        } else {
            this.getNextHandler().processRequest(request);
        }
    }
}

class Level3Handler extends IHandler {
    isAllowed(request) {
        return request.priority < 2;
    }

    processRequest(request) {
        if (this.isAllowed(request)) {
            console.log("Processing the request : lvel1 is processing");
        } else {
            this.getNextHandler().processRequest(request);
        }
    }
}

const level1 = new Level1Handler();
const level2 = new Level2Handler();
const level3 = new Level3Handler();

level1.setNextHandler(level2).setNextHandler(level3);

File: C++/DesignPatterns/ChainOfResponsibilityDesignPattern.cpp
#include <iostream>

struct Request {
    int priority = 0;
};

class IHandler {
public:
    virtual ~IHandler() = default;

    IHandler* setNextHandler(IHandler* handler) {
        nextHandler_ = handler;
        return nextHandler_;
    }

    IHandler* getNextHandler() const {
        return nextHandler_;
    }

    virtual bool isAllowed(const Request& request) = 0;
    virtual void processRequest(const Request& request) = 0;

private:
    IHandler* nextHandler_ = nullptr;
};

class Level1Handler : public IHandler {
public:
    bool isAllowed(const Request& request) override {
        return request.priority < 8;
    }

    void processRequest(const Request& request) override {
        if (isAllowed(request)) {
            std::cout << "Processing the request : lvel1 is processing" << std::endl;
        } else {
            getNextHandler()->processRequest(request);
        }
    }
};

class Level2Handler : public IHandler {
public:
    bool isAllowed(const Request& request) override {
        return request.priority < 15;
    }

    void processRequest(const Request& request) override {
        if (isAllowed(request)) {
            std::cout << "Processing the request : lvel1 is processing" << std::endl;
        } else {
            getNextHandler()->processRequest(request);
        }
    }
};

class Level3Handler : public IHandler {
public:
    bool isAllowed(const Request& request) override {
        return request.priority < 2;
    }

    void processRequest(const Request& request) override {
        if (isAllowed(request)) {
            std::cout << "Processing the request : lvel1 is processing" << std::endl;
        } else {
            getNextHandler()->processRequest(request);
        }
    }
};

int main() {
    Level1Handler level1;
    Level2Handler level2;
    Level3Handler level3;

    level1.setNextHandler(&level2)->setNextHandler(&level3);

    return 0;
}
